### PR TITLE
feat(string/location)

### LINKED ISSUE #77: Create string/location utility
from PR #76

### PR DESCRIPTION
**Pull request by AustinArey**
Closes #77 
- Added `locateRegex()` - this seems the most general way
- Anything else?

### CODE REVIEW

File: string/location.ts
Comment (pskfyi) [inline string/location.ts:3]: I think what we want is even simpler - just `location(text: string, index: number): TextLocation`. This can then be used in evalCodeBlocks but also anywhere else where we want to translate between an index (as given by regex match.index in this case, but there are other situations where this can be obtained) and a more sophisticated `TextLocation` description.

Comment (AustinArey) [reply string/location.ts:14]: Added `location()`

**Comment by AustinArey** [comment]
Ready

File: string/types.ts
Comment (pskfyi) [inline string/types.ts:2]: Since these are so closely connected to the functions, I think it's better to place them in that file with the functions. A standalone `types.ts` file is meant for shared types, as in if we had multiple functions within this module which returned these types or used them as inputs.

Once we have a `CONTRIBUTING.md` (#10) it would be good to locate guidance like this within the file.

File: string/location.ts
Comment (pskfyi) [inline string/location.ts:20]: This can be boiled down to:

```ts
return [...text.matchAll(regex)]
  .map((match) => [match[0], location(text, thisMatch.index ?? 0)])
```

This means that it's just as many lines to write out by hand as it is to import (1 line) and use (a 2nd line) this utility function, so it saves the user very little. It's closer to a snippet than a utility function... actually made me think we should consider creating snippets. Created #80 to track.

Anyway, I wouldn't recommend creating a util for this, but if you strongly disagree I won't block the PR over it 👍

File: string/location.ts
Comment (pskfyi) [inline string/location.ts:10]: We can return this immediately to remove a line without diminishing legibility. We can also use [object property shorthand syntax](https://eslint.org/docs/latest/rules/object-shorthand) to remove some noise.

```ts
return { offset: index, line, column }
```

File: string/location.ts
Comment (pskfyi) [inline string/location.ts:7]: Should we replace `linesText.length` with `line`?

**Review by pskfyi — COMMENTED** [review]
Looking great! Primarily it needs unit tests, a readme entry, and a JS Doc with an `@example`.

File: string/location.ts
Comment (AustinArey) [reply string/location.ts]: Agreed. Great point. Removed this and will consider a snippet later.

File: string/location.ts
Comment (AustinArey) [reply string/location.ts]: Agreed

File: readme.md
Comment (pskfyi) [inline readme.md:266]: It needs to be `export`ed from `string/utils.ts` for this to work 👍

**Comment by AustinArey** [comment]
Well I wish I snuck that in earlier. Thanks so much for your guidance on this!

**Review by pskfyi — APPROVED** [review]
LGTM!

**Comment by pskfyi** [comment]
Oh, sorry for the thrash on this but can we edit the commit message?

- Since this will show up in release notes, I'd like the `closes #77` comment to be in the commit message. This way the commit will always point to that ticket, which in turn will point to this PR.

- We should remove the duplicate reference to `location` in the scope (the part in parentheses)

- We should indicate that we're *adding* `location()` rather than some other `feat` such as adding new functionality to an already-existing `location()` function.

My recommendation in full:

> ``feat(string): add `location()`; closes #77``

**Comment by AustinArey** [comment]
> feat(string): add `location()`; closes #77

Thanks!

**Comment by pskfyi** [comment]
Thank you! Good to go when you are.

**Pull request merged**
